packages/create-remix/templates/fly-stack/app/routes/join.tsx

```diff
@@ -1,76 +1,138 @@
-import { Form, Link, redirect, useSearchParams } from "remix";
 import type { ActionFunction, LoaderFunction, MetaFunction } from "remix";
+import {
+  Form,
+  Link,
+  redirect,
+  useSearchParams,
+  json,
+  useActionData
+} from "remix";
+import Alert from "@reach/alert";
 
 import { getUserId, createUserSession } from "~/session.server";
 
 import { createUser } from "~/models/user.server";
-import invariant from "tiny-invariant";
 
-const loader: LoaderFunction = async ({ request }) => {
+export const loader: LoaderFunction = async ({ request }) => {
   const userId = await getUserId(request);
   if (userId) return redirect("/");
   return {};
 };
 
-const action: ActionFunction = async ({ request }) => {
+interface ActionData {
+  errors: {
+    email?: string;
+    password?: string;
+  };
+}
+
+export const action: ActionFunction = async ({ request }) => {
   const formData = await request.formData();
   const email = formData.get("email");
   const password = formData.get("password");
+  const returnTo = formData.get("returnTo");
+
+  if (typeof email !== "string" || email.length === 0) {
+    return json<ActionData>(
+      { errors: { email: "Email is required" } },
+      { status: 400 }
+    );
+  }
 
-  invariant(typeof email === "string", "email must be a string");
-  invariant(typeof password === "string", "password must be a string");
+  if (typeof password !== "string" || password.length === 0) {
+    return json<ActionData>(
+      { errors: { password: "Password is required" } },
+      { status: 400 }
+    );
+  }
 
   const user = await createUser(email, password);
 
   if (!user) {
-    return redirect("/login");
+    return json<ActionData>(
+      { errors: { email: "Invalid email or password" } },
+      { status: 400 }
```

Comment:
I think these should be [`422`s](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422) actually... @mjackson?

Reply:
the description of a [400](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400) fits better imo as it's "malformed request syntax"